### PR TITLE
feat: Improve DB performance by changing ES logic

### PR DESCRIPTION
**Pull request by Ali-D-Akbar**
[PROD-3873](https://2u-internal.atlassian.net/browse/PROD-3873)
Updates the underlying ES logic for AggregateSearchSerializer which uses CourseSearchDocumentSerializer and LearnerPathwaySearchDocumentSerializers. These serializers used to fetch each item from the DB one by one. This PR will stay backward compatible along with changing only the problematic serializers to fetch all objects in one DB call. 

This results in significant reduction of DB queryes of Course and LearnerPathway along with it reducing calls to their related models (courserun, courseruntype, seattype etc). Ultimately, this will greatly improve performance for the ES endpoints like `search/all/` and `search/courses/`.

### Testing Instruction

- Update ES index on local: `./manage.py update_index --disable-change-limit`
- Verify all the following URLs and check the decreased amount of queries:

```
router.register(r'search/limited', search_views.LimitedAggregateSearchView, basename='search-limited')
router.register(r'search/all', search_views.AggregateSearchViewSet, basename='search-all')
router.register(r'search/courses', search_views.CourseSearchViewSet, basename='search-courses')
router.register(r'search/course_runs', search_views.CourseRunSearchViewSet, basename='search-course_runs')
router.register(r'search/programs', search_views.ProgramSearchViewSet, basename='search-programs')
router.register(r'search/people', search_views.PersonSearchViewSet, basename='search-people')
```

### CODE REVIEW

File: course_discovery/apps/course_metadata/search_indexes/serializers/common.py [hunk 1]
```diff
@@ -25,26 +25,33 @@ class ModelObjectDocumentSerializerMixin:
     There is method `get_model_object_by_instance` to fetch Model object.
     """
 
-    def get_model_object_by_instance(self, instance):
+    def get_model_object_by_instance(self, instances):
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/search_indexes/serializers/common.py]: nit: I would suggest having a separate util for this, for sake of backwards compatibility.

File: course_discovery/apps/course_metadata/search_indexes/serializers/course.py [hunk 2]
```diff
@@ -120,15 +139,19 @@ def __init__(self, *args, **kwargs):
...
 
     def to_representation(self, instance):
-        _object = self.get_model_object_by_instance(instance)
-        setattr(instance, 'object', _object)  # pylint: disable=literal-used-as-attribute
-        return super().to_representation(instance)
+        if isinstance(instance, tuple):
+            setattr(instance[0], 'object', instance[1])  # pylint: disable=literal-used-as-attribute
+            prepared_instance = instance[0]
+        else:
+            prepared_instance = instance
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/search_indexes/serializers/course.py]: Dont we need to set object to instance in any case?

File: course_discovery/apps/edx_elasticsearch_dsl_extensions/serializers.py [hunk 3]
```diff
@@ -346,15 +351,19 @@ def patch_multi_serializer_context(self, serializer_class):
         serializer_class.Meta = Meta_origin
         serializer_class._declared_fields = origin_declared_fields  # pylint: disable=protected-access
 
-    def multi_serializer_representation(self, instance):
+    def multi_serializer_representation(self, instances):
```
Comment (DawoudSheraz) [inline course_discovery/apps/edx_elasticsearch_dsl_extensions/serializers.py]: Again, it would be better to have a new util method that represents what is being done here. This is no longer returning the representation, only serializers grouping.

File: course_discovery/apps/course_metadata/search_indexes/serializers/aggregation.py [hunk 4]
```diff
@@ -74,22 +76,51 @@ class Meta:
...
+        """
+        # Dealing with nested relationships, data can be a Manager,
+        # so, first get a queryset from the Manager if needed
+        iterable = data.all() if isinstance(data, models.Manager) else data
+        grouped_instances = self.multi_serializer_representation(iterable)
+
+        representation = []
+
+        for serializer_class, instances_for_serializer in dict(grouped_instances).items():
+            # Instantiate the serializer class
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/search_indexes/serializers/aggregation.py]: no need for this comment, it is obvious.

File: course_discovery/apps/course_metadata/search_indexes/serializers/aggregation.py [hunk 5]
```diff
@@ -74,22 +76,51 @@ class Meta:
...
 
-class AggregateSearchSerializer(MultiDocumentSerializerMixin, DocumentSerializer):
+class AggregateSearchListSerializer(MultiDocumentSerializerMixin, ListSerializer):
     """
-    Serializer for aggregated elasticsearch documents.
+    Custom List Serializer for AggregateSearch to fetch all instances at once.
     """
+    def to_representation(self, data):
+        """
+        List of object instances -> List of dicts of primitive datatypes.
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/search_indexes/serializers/aggregation.py]: nit: update docstring to include proper details of what is happening here.

File: course_discovery/apps/course_metadata/search_indexes/serializers/aggregation.py [hunk 6]
```diff
@@ -74,22 +76,51 @@ class Meta:
...
+        # so, first get a queryset from the Manager if needed
+        iterable = data.all() if isinstance(data, models.Manager) else data
+        grouped_instances = self.multi_serializer_representation(iterable)
+
+        representation = []
+
+        for serializer_class, instances_for_serializer in dict(grouped_instances).items():
+            # Instantiate the serializer class
+            representation += serializer_class(
+                context=self._context, many=True
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/search_indexes/serializers/aggregation.py:95]: do we really need many=True here, considering the individual serializers already have list_serializer_class defined?

File: course_discovery/apps/course_metadata/search_indexes/serializers/aggregation.py [hunk 7]
```diff
@@ -74,22 +76,51 @@ class Meta:
...
+    def to_representation(self, data):
+        """
+        List of object instances -> List of dicts of primitive datatypes.
+        """
+        # Dealing with nested relationships, data can be a Manager,
+        # so, first get a queryset from the Manager if needed
+        iterable = data.all() if isinstance(data, models.Manager) else data
+        grouped_instances = self.multi_serializer_representation(iterable)
+
+        representation = []
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/search_indexes/serializers/aggregation.py]: nit: representations

File: course_discovery/apps/course_metadata/search_indexes/serializers/course.py [hunk 8]
```diff
@@ -1,8 +1,11 @@
+# pylint: disable=W0223
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/search_indexes/serializers/course.py:1]: why is this needed?

File: course_discovery/apps/course_metadata/search_indexes/serializers/course.py [hunk 9]
```diff
@@ -18,7 +21,23 @@
...
 
 
-class CourseSearchDocumentSerializer(ModelObjectDocumentSerializerMixin, DateTimeSerializerMixin, DocumentSerializer):
+class CourseSearchDocumentListSerializer(ModelObjectDocumentSerializerMixin, ListSerializer):
+    """
+    Custom List Serializer for CourseSearchDocument to fetch all instances at once.
+    """
+    def to_representation(self, data):
+        """
+        List of object instances -> List of dicts of primitive datatypes.
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/search_indexes/serializers/course.py]: docstring update.

File: course_discovery/apps/course_metadata/search_indexes/serializers/program.py [hunk 10]
```diff
@@ -16,6 +19,20 @@
...
+    """
+    Custom List Serializer for ProgramSearchDocument to fetch all instances at once.
+    """
+    def to_representation(self, data):
+        """
+        List of object instances -> List of dicts of primitive datatypes.
+        """
+        # Dealing with nested relationships, data can be a Manager,
+        # so, first get a queryset from the Manager if needed
+        iterable = data.all() if isinstance(data, models.Manager) else data
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/search_indexes/serializers/program.py]: This to_representation is doing what DRF's default to_representation in ListSerializer does. It is not needed here it seems.

File: course_discovery/apps/edx_elasticsearch_dsl_extensions/serializers.py [hunk 11]
```diff
@@ -299,21 +299,26 @@ def to_representation(self, instance):
...
-            index_name = instance.meta['index']
-            index_alias = ElasticsearchUtils.get_alias_by_index_name(index_name)
-            serializer_class = instance_serializers.get(index_alias, None)
-            if not serializer_class:
-                raise ImproperlyConfigured('Could not find serializer for %s in mapping' % index_name)
-        return serializer_class
+    def get_serializer_classes_by_instances(self, instances):
+        if not isinstance(instances, list):
+            instances = [instances]
+        serializer_classes = []
```
Comment (DawoudSheraz) [inline course_discovery/apps/edx_elasticsearch_dsl_extensions/serializers.py]: Add a comment explaining the situation where `instances` might not be a list.

File: course_discovery/apps/course_metadata/search_indexes/serializers/common.py (hunk 1 above)
Comment (Ali-D-Akbar) [reply course_discovery/apps/course_metadata/search_indexes/serializers/common.py]: I have made the util backwards compatible by adding this

```
        if not isinstance(instances, list):
            instances = [instances]
```

Comment (Ali-D-Akbar) [reply course_discovery/apps/course_metadata/search_indexes/serializers/common.py]: I'll update the name for this.

File: course_discovery/apps/course_metadata/search_indexes/serializers/course.py (hunk 2 above)
Comment (Ali-D-Akbar) [reply course_discovery/apps/course_metadata/search_indexes/serializers/course.py]: Just in case the custom list serializer isn't invoked, we won't have an instance with object so it handles that case.

File: course_discovery/apps/course_metadata/search_indexes/serializers/aggregation.py (hunk 6 above)
Comment (Ali-D-Akbar) [reply course_discovery/apps/course_metadata/search_indexes/serializers/aggregation.py:95]: In this instance, yes. This will invoke the respective custom list serializers for all the documents.

File: course_discovery/apps/course_metadata/search_indexes/serializers/course.py (hunk 8 above)
Comment (Ali-D-Akbar) [reply course_discovery/apps/course_metadata/search_indexes/serializers/course.py:1]: Pylint fails. Says`serializer_class` is not callable.

File: course_discovery/apps/course_metadata/search_indexes/serializers/program.py (hunk 10 above)
Comment (Ali-D-Akbar) [reply course_discovery/apps/course_metadata/search_indexes/serializers/program.py]: No, it is needed as well. It sends the whole queryset rather than iterating over it.

File: course_discovery/apps/edx_elasticsearch_dsl_extensions/serializers.py (hunk 11 above)
Comment (Ali-D-Akbar) [reply course_discovery/apps/edx_elasticsearch_dsl_extensions/serializers.py]: Nevermind I've separated the util functions for them.

File: course_discovery/apps/course_metadata/search_indexes/serializers/aggregation.py (hunk 6 above)
Comment (DawoudSheraz) [reply course_discovery/apps/course_metadata/search_indexes/serializers/aggregation.py:95]: But the docserializers already have list_serializer_class defined, shouldn't that do the job? For instance, the same approach is used for AggregateSearch serializer and it works with many=true when called.

File: course_discovery/apps/course_metadata/search_indexes/serializers/course.py (hunk 2 above)
Comment (DawoudSheraz) [reply course_discovery/apps/course_metadata/search_indexes/serializers/course.py]: Right, but don't we need to do this in "else"?
```
_object = self.get_model_object_by_instance(instance)
  setattr(instance, 'object', _object)
```

File: course_discovery/apps/course_metadata/search_indexes/serializers/aggregation.py [hunk 12]
```diff
@@ -74,22 +76,51 @@ class Meta:
...
-    Serializer for aggregated elasticsearch documents.
+    Custom List Serializer for AggregateSearch to fetch all instances at once.
     """
+    def to_representation(self, data):
+        """
+        Custom representation for all the grouped multi-serializer instances.
+        This will invoke the respective custom list serializers for all the documents.
+        """
+        # Dealing with nested relationships, data can be a Manager,
+        # so, first get a queryset from the Manager if needed
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/search_indexes/serializers/aggregation.py]: no need for this comment as it is a default comment.

File: course_discovery/apps/course_metadata/search_indexes/serializers/common.py [hunk 13]
```diff
@@ -69,5 +77,5 @@ class DocumentDSLSerializerMixin(ModelObjectDocumentSerializerMixin):
     """
 
     def to_representation(self, instance):
-        _object = self.get_model_object_by_instance(instance)
+        _object = self.get_model_object_by_instances(instance).get()
         return super().to_representation(_object)
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/search_indexes/serializers/common.py:81]: Since this is in common.py, don't we need to handle tuple usecase here as well?

File: course_discovery/apps/edx_elasticsearch_dsl_extensions/serializers.py [hunk 14]
```diff
@@ -358,3 +383,15 @@ def multi_serializer_representation(self, instance):
             representation = serializer_class(context=self._context).to_representation(instance)
 
         return representation
+
+    def group_multi_serializer_instances(self, instances):
+        """
+        Groups Multi serializer instances in a list of dicts.
+        """
+        serializer_classes = self.get_serializer_classes_by_instances(instances)
```
Comment (DawoudSheraz) [inline course_discovery/apps/edx_elasticsearch_dsl_extensions/serializers.py]: Can't this be consolidated in here? We are getting serializers list first and then looping over it again via zip (to create dict mapping).

File: course_discovery/apps/course_metadata/search_indexes/serializers/aggregation.py (hunk 6 above)
Comment (Ali-D-Akbar) [reply course_discovery/apps/course_metadata/search_indexes/serializers/aggregation.py:95]: This actually invokes the custom serializers in `list_serializer_class` which means each serializer will handle it's own serializing.

File: course_discovery/apps/course_metadata/search_indexes/serializers/course.py (hunk 2 above)
Comment (Ali-D-Akbar) [reply course_discovery/apps/course_metadata/search_indexes/serializers/course.py]: We're not fetching objects from DB in case of programs and courseruns. I've added this custom serializer to keep consistency with the **Grouped Serializers** approach.

Comment (DawoudSheraz) [reply course_discovery/apps/course_metadata/search_indexes/serializers/course.py]: But this is course.py, not courseruns.

File: course_discovery/apps/course_metadata/search_indexes/serializers/program.py (hunk 10 above)
Comment (DawoudSheraz) [reply course_discovery/apps/course_metadata/search_indexes/serializers/program.py]: But the queryset is being passed to super().to_representation, which essentially does the same iteration over queryset.

File: course_discovery/apps/course_metadata/search_indexes/serializers/course.py (hunk 2 above)
Comment (Ali-D-Akbar) [reply course_discovery/apps/course_metadata/search_indexes/serializers/course.py]: Adding this in the instance as discussed

File: course_discovery/apps/course_metadata/search_indexes/serializers/program.py (hunk 10 above)
Comment (Ali-D-Akbar) [reply course_discovery/apps/course_metadata/search_indexes/serializers/program.py]: Ah right. It makes sense. Removing them for program, courserun and person docs.

File: course_discovery/apps/course_metadata/search_indexes/serializers/common.py (hunk 13 above)
Comment (Ali-D-Akbar) [reply course_discovery/apps/course_metadata/search_indexes/serializers/common.py:81]: I'm keeping the behavior intact for DSL search variants of documents like `CourseSearchModelSerializer`, `CourseRunSearchModelSerializer` bcz I honestly don't know how `DocumentDSLSerializerMixin` work and don't want to disrupt their working/performance.

File: course_discovery/apps/edx_elasticsearch_dsl_extensions/serializers.py (hunk 14 above)
Comment (Ali-D-Akbar) [reply course_discovery/apps/edx_elasticsearch_dsl_extensions/serializers.py]: Right. I'm on it.

File: course_discovery/apps/course_metadata/search_indexes/serializers/aggregation.py [hunk 15]
```diff
@@ -74,22 +76,49 @@ class Meta:
...
+        """
+        Custom representation for all the grouped multi-serializer instances.
+        This will invoke the respective custom list serializers for all the documents.
+        """
+        iterable = data.all() if isinstance(data, models.Manager) else data
+        grouped_instances = self.group_multi_serializer_instances(iterable)
+
+        representations = []
+
+        for serializer_class, instances_for_serializer in dict(grouped_instances).items():
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/search_indexes/serializers/aggregation.py]: grouped_instances is already a dict, so why call dict() on it again?

File: course_discovery/apps/course_metadata/search_indexes/serializers/common.py [hunk 16]
```diff
@@ -69,5 +77,5 @@ class DocumentDSLSerializerMixin(ModelObjectDocumentSerializerMixin):
     """
 
     def to_representation(self, instance):
-        _object = self.get_model_object_by_instance(instance)
+        _object = self.get_model_object_by_instances(instance).get()
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/search_indexes/serializers/common.py:80]: nit: don't we need to set the object to instance?

File: course_discovery/apps/course_metadata/search_indexes/serializers/common.py (hunk 13 above)
Comment (DawoudSheraz) [reply course_discovery/apps/course_metadata/search_indexes/serializers/common.py:81]: It is better to make the code defensive then.

File: course_discovery/apps/course_metadata/search_indexes/serializers/course.py [hunk 17]
```diff
@@ -18,6 +21,20 @@
 __all__ = ('CourseSearchDocumentSerializer',)
 
 
+class CourseSearchDocumentListSerializer(ModelObjectDocumentSerializerMixin, ListSerializer):
+    """
+    Custom List Serializer for CourseSearchDocument to fetch all instances at once.
+    """
+    def to_representation(self, data):
+        """
+        Custom representation for all the course instances.
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/search_indexes/serializers/course.py]: docstring should specify why do we need this custom representation.

File: course_discovery/apps/course_metadata/search_indexes/serializers/course.py [hunk 18]
```diff
@@ -18,6 +21,20 @@
...
+        """
+        iterable = data.all() if isinstance(data, models.Manager) else data
+        _objects = self.get_model_object_by_instances(iterable)
+        objects_list = list(_objects)
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/search_indexes/serializers/course.py]: nit: both can be combined

File: course_discovery/apps/course_metadata/search_indexes/serializers/course.py [hunk 19]
```diff
@@ -120,15 +137,21 @@ def __init__(self, *args, **kwargs):
             self.fields.pop('outcome', None)
 
     def to_representation(self, instance):
-        _object = self.get_model_object_by_instance(instance)
-        setattr(instance, 'object', _object)  # pylint: disable=literal-used-as-attribute
-        return super().to_representation(instance)
+        if isinstance(instance, tuple):
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/search_indexes/serializers/course.py]: update docstring to differentiate b/w instance being a tuple and instance being a Hit.

File: course_discovery/apps/course_metadata/search_indexes/serializers/learner_pathway.py [hunk 20]
```diff
@@ -12,6 +15,17 @@
 __all__ = ('LearnerPathwaySearchDocumentSerializer',)
 
 
+class LearnerPathwaySearchDocumentListSerializer(ModelObjectDocumentSerializerMixin, ListSerializer):
+    def to_representation(self, data):
+        """
+        Custom representation for all the learner_pathway instances.
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/search_indexes/serializers/learner_pathway.py]: same docstring comment as above to be applied here.

File: course_discovery/apps/course_metadata/search_indexes/serializers/common.py [hunk 21]
```diff
@@ -22,29 +22,37 @@ class ModelObjectDocumentSerializerMixin:
...
-        Provide Model object by elasticsearch response instance.
+        Provide Model objects by elasticsearch response instances.
+        Fetches all the incoming instances at once and returns model queryset.
         """
+        if not isinstance(instances, list):
+            instances = [instances]
         document = None
-        _object = None
-        index_or_alias_name = ElasticsearchUtils.get_alias_by_index_name(instance.meta.index)
+        _object = []
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/search_indexes/serializers/common.py]: nit: `_objects`

**Review by DawoudSheraz — COMMENTED** [review]
Tested locally. There are two N+1 issues:
1. additional_metadata on course
2. Various models on LearnerPathway
Otherwise, the code seems to be working. I have already left some nit/changes that are to be catered before merge.

File: course_discovery/apps/course_metadata/search_indexes/serializers/aggregation.py (hunk 15 above)
Comment (Ali-D-Akbar) [reply course_discovery/apps/course_metadata/search_indexes/serializers/aggregation.py]: Updated.

File: course_discovery/apps/course_metadata/search_indexes/serializers/common.py (hunk 16 above)
Comment (Ali-D-Akbar) [reply course_discovery/apps/course_metadata/search_indexes/serializers/common.py:80]: Not in this case for some reason 🤔

File: course_discovery/apps/edx_elasticsearch_dsl_extensions/serializers.py [hunk 22]
```diff
@@ -358,3 +361,14 @@ def multi_serializer_representation(self, instance):
             representation = serializer_class(context=self._context).to_representation(instance)
 
         return representation
+
+    def group_multi_serializer_instances(self, instances):
+        """
+        Groups Multi serializer instances in a list of dicts.
+        """
+        grouped_instances = defaultdict(list)
```
Comment (AfaqShuaib09) [inline course_discovery/apps/edx_elasticsearch_dsl_extensions/serializers.py:369]: question: any reason behind using defaultdict here?

**Review by AfaqShuaib09 — APPROVED** [review]

File: course_discovery/apps/course_metadata/search_indexes/serializers/course.py [hunk 23]
```diff
@@ -18,6 +21,19 @@
...
+    """
+    Custom List Serializer for CourseSearchDocument to fetch all instances at once.
+    """
+    def to_representation(self, data):
+        """
+        Custom list representation to fetch all the course instances at once.
+        """
+        iterable = data.all() if isinstance(data, models.Manager) else data
+        _objects = list(self.get_model_object_by_instances(iterable))
+        return super().to_representation(list(zip(iterable, _objects)))
```
Comment (zawan-ila) [inline course_discovery/apps/course_metadata/search_indexes/serializers/course.py]: I do not think this is correct. You can not assume that the order of `_objects` will match that of iterable. Hence you'll end up zipping wrong objects. i.e you'll end up zipping Course1ESObject with Course2SQLObject and when it will be serialized, you'll have mixed i.e incorrect info.

Comment (zawan-ila) [reply course_discovery/apps/course_metadata/search_indexes/serializers/course.py]: Can you verify if you get the correct results here?

Comment (zawan-ila) [reply course_discovery/apps/course_metadata/search_indexes/serializers/course.py]: To illustrate my point, here's an object I get in the output of `/search/all`
Screenshot 2024-01-22 at 5 37 01 PM
Note How the ES info and the SQL info belong to different courses.

*I hope it's not just because my local is messed up :-)

Comment (zawan-ila) [reply course_discovery/apps/course_metadata/search_indexes/serializers/course.py]: Also you seem to be doing similar zip for `learner_pathway`. You'll need to fix it there too.

Comment (DawoudSheraz) [reply course_discovery/apps/course_metadata/search_indexes/serializers/course.py]: Guess this will need to be handled manually via iteration then(?). Both queryset and instances are lists, so the code would need to search for each item individually (whether the hits or queryset is picked). 
However, that brings the question, of the primary keys that are provided to model via list. That list is maintained via the order of instances. That should bring queryset in same order as the provided keys.

Comment (DawoudSheraz) [reply course_discovery/apps/course_metadata/search_indexes/serializers/course.py]: If the queryset is indeed in the order of pks provided, this should work out of box (provided the item length is same)
```
>>> Course.everything.filter(id__in=([1, 50, 56, 90, 100])).values_list('pk')
<CourseQuerySet [(1,), (50,), (56,), (90,), (100,)]>
```

Comment (DawoudSheraz) [reply course_discovery/apps/course_metadata/search_indexes/serializers/course.py]: Ok, never mind. I verified it locally, this does mess up the results.
image

Comment (DawoudSheraz) [reply course_discovery/apps/course_metadata/search_indexes/serializers/course.py]: The ordering messes up because the hit instances are not ordered by pks. When the pks list is sent to queryset, it returns the items ordered by pks (by default). This causes an incorrect tuple setting in zip().

File: course_discovery/apps/edx_elasticsearch_dsl_extensions/serializers.py (hunk 22 above)
Comment (Ali-D-Akbar) [reply course_discovery/apps/edx_elasticsearch_dsl_extensions/serializers.py:369]: Just to define an empty dict.

File: course_discovery/apps/course_metadata/search_indexes/serializers/course.py (hunk 23 above)
Comment (Ali-D-Akbar) [reply course_discovery/apps/course_metadata/search_indexes/serializers/course.py]: Added `iterable.sort(key=lambda x: x.pk) if iterable else None`

File: course_discovery/apps/course_metadata/search_indexes/serializers/course.py [hunk 24]
```diff
@@ -120,15 +141,28 @@ def __init__(self, *args, **kwargs):
...
     def to_representation(self, instance):
-        _object = self.get_model_object_by_instance(instance)
-        setattr(instance, 'object', _object)  # pylint: disable=literal-used-as-attribute
-        return super().to_representation(instance)
+        """
+        Custom instance representation.
+        The instance needs to be handled differently and can be either of the two:
+
+        1. A tuple consistent of a Hit object and a model object to be assigned to the hit object.
+        2. A single model object
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/search_indexes/serializers/course.py]: 2. Isn't it Hit/doc instance, against which the DB object is fetched?

Comment (Ali-D-Akbar) [reply course_discovery/apps/course_metadata/search_indexes/serializers/course.py]: Oh.. wait.

File: course_discovery/apps/course_metadata/search_indexes/serializers/common.py (hunk 16 above)
Comment (DawoudSheraz) [reply course_discovery/apps/course_metadata/search_indexes/serializers/common.py:80]: What is that "some reason" then?

Comment (Ali-D-Akbar) [reply course_discovery/apps/course_metadata/search_indexes/serializers/common.py:80]: (Found the reason after this comment) `DocumentDSLSerializerMixin` is used in `CourseSearchModelSerializer` which creates document from model only. (without using index)

File: course_discovery/apps/course_metadata/search_indexes/serializers/learner_pathway.py [hunk 25]
```diff
@@ -44,9 +62,21 @@ class Meta:
...
     def to_representation(self, instance):
-        _object = self.get_model_object_by_instance(instance)
-        setattr(instance, 'object', _object)  # pylint: disable=literal-used-as-attribute
-        return super().to_representation(instance)
+        """
+        Custom instance representation.
+        The instance needs to be handled differently and can be either of the two:
+
+        1. A tuple consistent of a Hit object and a model object to be assigned to the hit object.
+        2. A single model object
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/search_indexes/serializers/learner_pathway.py]: nit: 2 is single hit object.

**Review by zawan-ila — APPROVED** [review]
